### ISSUE TITLE
No PCR0002 on missing await on `.ConfigureAwait(bool)`

### ISSUE THREAD

**Issue by ramonsmits** [Particular/Particular.Analyzers#30]
I expected the following code to raise a PCR0002 error but it compiles without any issue.

```c#
static void Expect_PCR0002(IEndpointInstance endpointInstance)
{
    // Expression creates a Task that is not awaited or assigned to a variable.
    Test().ConfigureAwait(false);
}
```

**Comment by danielmarbach** [comment]
From the perspective of the analyzer the task is consumed and this is valid code (although doesn't achieve what people probably intended)

**Comment by timbussmann** [comment]
Code Rules is a set of analyzers used internally btw.
I think what you're looking for is [NSB0001](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core.Analyzer/AwaitOrCaptureTasksAnalyzer.cs) in the core repo?

The core analyzer is limitted to a specific set of public APIs to check. ConfigureAwait makes it much more difficult to find the relevant use cases so it was decided back then to not support ConfigureAwait to start with.

**Comment by ramonsmits** [comment]
@danielmarbach PCR0002 is generic and not inspects only our types. As `ConfigureAwait(bool)` also returns a task which is not assigned this seems like a case that could be checked for.

@timbussmann This package is also used by customers. If this should be used by customers we shouldn't have put it on Nuget.

I understand the reasoning as that is probably the same reason VS is not showing a warning/error.

I'm  raising it here as a use case that fails the check. This either this must be fixed or documented.

**Comment by timbussmann** [comment]
> This either this must be fixed or documented.

this is not an officially supported package (https://docs.particular.net/nservicebus/upgrades/supported-versions) and it's purpose is for internal use. I don't think there is something to "fix" here, if at all we could improve the analyzer in case there is an internal need for it. For analyzers issues related to customers, we should talk about the mentioned core analyzer and not this one.

> This package is also used by customers. If this should be used by customers we shouldn't have put it on Nuget.

there are plenty of nugget packages on nugget that aren't officially supported and shouldn't be used. That's a different story (and I agree that those shouldn't be there).

btw. we should probably look into whether this analyzer is required at all since VS has it's own async analyzers by now.

**Comment by adamralph** [comment]
This diagnostic is now PS0001, and no, it does not report violations when `.ConfigureAwait(false)` is present.

Incidentally, PS0001 does appear to be redundant, since this is now covered by [CS4014](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs4014):

![image](https://user-images.githubusercontent.com/677704/115410686-208a6e00-a1f3-11eb-8db2-85ee4a1b062f.png)

Note that CS4014 also does not report violations when `.ConfigureAwait(false)` is present:

![image](https://user-images.githubusercontent.com/677704/115410853-4879d180-a1f3-11eb-903a-9e093fde226f.png)

**Comment by timbussmann** [comment]
> This diagnostic is now PS0001

@adamralph  What does that mean? Can you give more context about that change? PS0001 was not about ConfigureAwait.

IIRC CS4014 doesn't work when the method is not marked as `async`, has that been changed?

**Comment by adamralph** [comment]
@timbussmann

> What does that mean? Can you give more context about that change? PS0001 was not about ConfigureAwait.

In Particular.Analyzers 1.0.0, PCR0002 was replaced with PS0001. When we changed the name of the package to Particular.Analyzers, we also changed from using PCR to PS as the diagnostic ID prefix, and we also reset the ID's to start at 1 (PCR0001 was removed in some old version of Particular.CodeRules).

> IIRC CS4014 doesn't work when the method is not marked as `async`, has that been changed?

It seems it does work even when the method does not have the `async` modifier:

<img src="https://user-images.githubusercontent.com/677704/115448761-60b11700-a21a-11eb-9d3c-12a11018f21b.png" width="415px" />

However, bizarrely, it doesn't show on the `Task.Delay` call:

<img src="https://user-images.githubusercontent.com/677704/115448797-6ad31580-a21a-11eb-84ca-7a89eb5561e9.png" width="375px" />

So I guess we still need PS0001 to be sure. 🤷

**Comment by ramonsmits** [comment]
Why is this closed? The problem as stated in the title is still present. Was a NO FIX decision made?

**Comment by adamralph** [comment]
replaced by https://github.com/Particular/Particular.Analyzers/pull/96
